File: Forge/src/main/java/hantonik/atomic/core/registration/object/FluidObject.java
package hantonik.atomic.core.registration.object;

import lombok.Getter;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.tags.FluidTags;
import net.minecraft.tags.TagKey;
import net.minecraft.world.item.Item;
import net.minecraft.world.level.ItemLike;
import net.minecraft.world.level.block.LiquidBlock;
import net.minecraft.world.level.material.Fluid;
import net.minecraftforge.fluids.ForgeFlowingFluid;
import net.minecraftforge.registries.ForgeRegistries;
import org.jetbrains.annotations.Nullable;

import javax.annotation.Nonnull;
import java.util.Objects;
import java.util.function.Supplier;

public class FluidObject<F extends ForgeFlowingFluid> implements Supplier<F>, ItemLike {
    protected ResourceLocation id;

    @Nullable
    private final Supplier<? extends LiquidBlock> block;
    private final Supplier<? extends F> still;
    private final Supplier<? extends F> flowing;

    @Getter
    private TagKey<Fluid> localTag;
    @Getter
    private TagKey<Fluid> forgeTag;

    public FluidObject(ResourceLocation id, String tagName, Supplier<? extends F> still, Supplier<? extends F> flowing, @Nullable Supplier<? extends LiquidBlock> block) {
        this.id = id;

        this.block = block;
        this.still = still;
        this.flowing = flowing;

        this.localTag = FluidTags.create(id);
        this.localTag = FluidTags.create(new ResourceLocation("forge", tagName));
    }

    @Deprecated
    public FluidObject(Supplier<? extends F> still, Supplier<? extends F> flowing, @Nullable Supplier<? extends LiquidBlock> block) {
        this.block = block;
        this.still = still;
        this.flowing = flowing;

        this.localTag = FluidTags.create(this.getId());
        this.forgeTag = FluidTags.create(new ResourceLocation("forge", this.getId().getPath()));
    }

    public ResourceLocation getId() {
        if (this.id == null)
            this.id = Objects.requireNonNull(ForgeRegistries.FLUIDS.getKey(this.getStill()), "Fluid has null ID");

        return this.id;
    }

    @Nullable
    public LiquidBlock getBlock() {
        return this.block == null ? null : this.block.get();
    }

    public F getStill() {
        return Objects.requireNonNull(this.still.get(), "Fluid object missing still fluid");
    }

    public F getFlowing() {
        return Objects.requireNonNull(this.flowing.get(), "Fluid object missing flowing fluid");
    }

    @Override
    public F get() {
        return this.getStill();
    }

    @Nonnull
    public Item asItem() {
        return this.still.get().getBucket();
    }
}

File: Forge/src/main/kotlin/hantonik/atomic/core/registration/object/FluidObject.kt
package hantonik.atomic.core.registration.`object`

import net.minecraft.resources.ResourceLocation
import net.minecraft.tags.FluidTags
import net.minecraft.tags.TagKey
import net.minecraft.world.item.Item
import net.minecraft.world.level.ItemLike
import net.minecraft.world.level.block.LiquidBlock
import net.minecraft.world.level.material.Fluid
import net.minecraftforge.fluids.ForgeFlowingFluid

open class FluidObject<F : ForgeFlowingFluid>(val id: ResourceLocation, tagName: String, private val still: () -> F?, private val flowing: () -> F?, val block: (() -> LiquidBlock)?) : () -> F, ItemLike {
    val localTag: TagKey<Fluid> = FluidTags.create(this.id)
    val forgeTag: TagKey<Fluid> = FluidTags.create(ResourceLocation("forge", tagName))

    fun getStill(): F {
        this.still.invoke() ?: throw NullPointerException("Fluid object missing still fluid")

        return this.still.invoke()!!
    }

    fun getFlowing(): F {
        this.flowing.invoke() ?: throw NullPointerException("Fluid object missing flowing fluid")

        return this.flowing.invoke()!!
    }

    override fun invoke(): F = this.getStill()

    override fun asItem(): Item = this.getStill().bucket
}